cs/src/Connections/TunnelConnection.cs

```diff
@@ -78,7 +79,32 @@ protected set
     /// Get the tunnel that is being hosted or connected to.
     /// May be null if the tunnel client used relay service URL and tunnel access token directly.
     /// </summary>
-    public Tunnel? Tunnel { get; private set; }
+    public Tunnel? Tunnel {
+        get => this.tunnel;
+        private set
+        {
+            if (value != this.tunnel)
+            {
+                // Get the tunnel access token from the new tunnel, or the original Tunnal object if the new tunnel doesn't have the token,
+                // which may happen when the tunnel was authenticated with a tunnel access token from Tunnel.AccessTokens.
+                // Add the tunnel access token to the new tunnel's AccessTokens if it is not there.
```

Comment:
I was planning to address this in the management APIs. See https://github.com/microsoft/basis-planning/issues/990